server/src/main/java/io/crate/types/DataType.java

```diff
@@ -288,4 +289,8 @@ public long ramBytesUsed() {
      */
     public void addMappingOptions(Map<String, Object> mapping) {
     }
+
+    public ColumnStatsSupport<T> columnStatsSupport() {
+        throw new UnsupportedOperationException("Datatype " + this + " does not support column stats");
+    }
```

Comment:
It looks like the unsupported case is not handled in https://github.com/crate/crate/blob/5b907d4e115026739be0476cd1ebe9a0e7b05976/server/src/main/java/io/crate/statistics/ReservoirSampler.java#L155

Makes me wonder if we should move this into `StorageSupport` and force the implementation instead of having a default failure case. Because each column we can store should have the capability to generate stats (even if it's just empty, like it might be the case for objects?).
If we can't store a type there is no need to implement stats building.

I'd also consider skipping the intermediate `ColumnStatsSupport` and expose the sketch builder directly.

On that note: I'm a bit confused by the current split into ColumnSketch/ColumnSketchBuilder/ColumnStats. It's not clear to me why we need to further distinguish between ColumnSketch and ColumnSketchBuilder. I'd assume we only need 2 classes instead of 3.

Couldn't we have:

- `ColumnStats.Builder` used in the ReservoirSampler and streamed across nodes
    - Uses `UpdateSketch` or whatever is needed from the sketch library
    - Used throughout the stats collection, and can be merged
    - Streamable via a `ColumnStats.Builder(StreamInput in)` (and type parameter if required)
    - Has a `ColumnStats.Builder.build()` -> returns the final `ColumnStats`
    - `ColumnStats` is no longer streamable. `ColumnSkets` (as it is in this PR) and `ColumnStats` are unified/they're the same thing?


(Didn't look at the rest in detail yet, as I suspect this would have a bit of an impact)

Reply:
I'll have a look at moving things directly into StorageSupport.

Once a sketch has been merged it can't be updated any more, so there's a fundamental difference between the builder, which accepts new values, and the final sketch, which is immutable.  From a code perspective, the type of the distinct sketch within `ColumnSketchBuilder` is an `UpdateSketch`, but the result of merging two of them is an immutable anonymous subclass of `Sketch`.  Rather than having something which is immutable up to a point but then frozen, and has different member fields before and after merging, it made sense to me to split it into two different classes.

Reply:
And `ColumnSketch` and `ColumnStats` need to remain separate because `ColumnSketch` needs to hold on to the sketch datastructures for merging/streaming, but there's a bit of processing to be done to generate the `ColumnStats` from it and it makes sense to just do that once I think?  We may be able to remove `Writeable` from `ColumnStats` though, I'll take a look.

Reply:
Ah no, `ColumnStats` needs to be `Writeable` as its used in `PublishTableStatsRequest`

Reply:
> from a code perspective, the type of the distinct sketch within ColumnSketchBuilder is an UpdateSketch, but the result of merging two of them is an immutable anonymous subclass of Sketch. Rather than having something which is immutable up to a point but then frozen, and has different member fields before and after merging, it made sense to me to split it into two different classes.

Is this a limitation of the sketch algorithm, or where is this coming from? I'd have expected that it's possible to further add data to a merged sketch, and also merge more sketches into a already merged sketch?


> And ColumnSketch and ColumnStats need to remain separate because ColumnSketch needs to hold on to the sketch datastructures for merging/streaming,

> Ah no, ColumnStats needs to be Writeable as its used in PublishTableStatsRequest


That's why my suggestion was to keep using the `ColumnStats.Builder` (or ColumnSketchBuilder) throughout, until the very end. Then we create the stats, when we no longer need the sketch datastructures for merging.

Reply:
> Is this a limitation of the sketch algorithm, or where is this coming from? 

It looks as though it's a limitation of the theta sketch algorithm - the sketch gets compacted for streaming and/or merging, and can't be updated afterwards.  There are other distinct sketch types in the library, so I'll have a look and see if any of them behave differently here.